src/animation.cpp

```diff
@@ -862,6 +862,46 @@ void game::draw_zones( const tripoint &start, const tripoint &end, const tripoin
 }
 #endif
 
+#if defined(TILES)
+void game::draw_async_anim( const tripoint &p, const std::string tile_id, const std::string ncstr,
+                            const nc_color nccol )
+{
+    if( test_mode ) {
+        // avoid segfault from null tilecontext in tests
+        return;
+    }
+
+    if( !get_option<bool>( "ANIMATIONS" ) ) {
+        return;
+    }
+
+    if( !u.sees( p ) ) {
+        return;
+    }
+
+    if( !use_tiles ) {
+        if( ncstr != "" ) {
+            hit_animation( get_avatar(), p, nccol, ncstr );
+        }
+        return;
+    }
+
+    shared_ptr_fast<draw_callback_t> async_anim_cb = make_shared_fast<draw_callback_t>( [&]() {
+        tilecontext->init_draw_async_anim( p, tile_id );
+    } );
+    add_draw_callback( async_anim_cb );
+    ui_manager::redraw();
+}
+#else
+void game::draw_async_anim( const tripoint &p, const std::string, const std::string ncstr,
+                            const nc_color nccol )
+{
+    if( ncstr != "" ) {
+        hit_animation( get_avatar(), p, nccol, ncstr );
```

Comment:
Does this mean it will still wait for the animation to finish in curses?

Reply:
Nope.  It will cancel any running animations when an input is received.

Reply:
Well, you're right it will, but it will also ignore the first input when the animation is running:

https://github.com/CleverRaven/Cataclysm-DDA/blob/9e4ff331dfed15b45abd55012951dcc847c36b44/src/animation.cpp#L547-L550

It should probably be made consistent with tiles (cancel the animation and handle the input action)

Reply:
Apparently, handling the input by passing an `action_id` from `look_up_action( handle_input() )` to `game::handle_action( action_id act )` leads to a bunch of issues like actions not taking any time.  I'm not exactly familiar with the input loop, but I guess it has something to do with handling actions outside of do_turn().  I don't suppose you know a better way to go about this?

Reply:
I think you can store and use the animation data in curses similar to how `init_draw_async_anim` does on tiles. `add_redraw_callback` may be used but you'll have to store the `shared_ptr`s somewhere because `class game` only stores weak pointers to the callbacks.

Reply:
Done.  I skipped the callbacks bit and just drew from an "animation layer" var like how it is done with tiles.  Works good so far on no-tiles mode.  Will test on real curses once I have my virtual machine handy.

Reply:
Working now.  All that is left is the serialization warning for `async_anim_layer_curses`.  I don't intend to store it in the save game.  Not sure if there is a better place to store it than `class game` though.

Reply:
You can add `//NOLINT(cata-serailize)` after the line to suppress the warning.

Reply:
Thanks.  Should have noticed how this is a standard practice.